### ISSUE TITLE
Discussion: Redesign server to use asynchronous DB system

### ISSUE THREAD

**Issue by mgaree** [JackieChiles/Cinch#115]
Current design is (trying to use) an active server program that stores game state information, provides a web server interface for clients, and facilitates AIs, as well as everything else the game does. When the server is terminated, all game info is lost, except for what was logged earlier.

I propose for discussion a new design, using a proper database to store game state information and such. Instead of clients querying a homemade Comet server, they query a URL that checks the database for new information. When clients make moves, they POST to a URL that applies the move to a game that exists in the DB rather than in memory. Except for maybe a persistent program that manages AI agents, the main game engine would be on-demand.

Some benefits:
- mySQL is designed for multiple simultaneous connections, so multiple client access should be a nonissue (not sure about SQLite)
- game data is always saved, so temporary disruption to the server doesn't cause the game to die for the clients
  -- storing session cookies could easily add to client stability
- synergy with current use of DB for logging
- no messy Comet server to support -- this is huge
- awkward command consoles not needed to interface with the program

Some drawbacks:
- DB access may be slow; would need to do profiling/benchmarking to know more
  -- each DB read would have to create some sort of game object to act on
- requires complete rewrite of server code, and some rewrite to client code
  -- probably only game/meta-game rules and AI logic could be preserved
- would probably still need active program for AI management

Bottom line: Cinch is ready to resume development, either along the path we left it on five months ago, or on a new path. A new design would allow us to repair our poor decisions made oh so long ago (while maybe not making some new ones). A new design would also add significant delay, effectively restarting an almost-there product. But I'm not casting a vote yet. Let's discuss.

Recommend that this issue be closed following discussion and consensus on course of action no later than 4 July 2013.
